pkg/eestream/encode.go

```diff
@@ -202,7 +204,8 @@ func (er *encodedReader) fillBuffer() {
 func (er *encodedReader) copyData(num int, copier <-chan block) {
 	// close the respective buffer channel when this goroutine exits
 	defer func() {
-		if er.eps[num].ch != nil {
+		if atomic.LoadInt32(er.eps[num].closed) == 0 {
+			atomic.StoreInt32(er.eps[num].closed, 1)
```

Comment:
This still has a race, at this point the closed status may already have changed.

Reply:
``` 
// goroutine 1                                      // goroutine 2

if atomic.LoadInt32(er.eps[num].closed) == 0 {      
                                                    if atomic.LoadInt32(er.eps[num].closed) == 0 {
                                                        atomic.StoreInt32(er.eps[num].closed, 1)
    atomic.StoreInt32(er.eps[num].closed, 1)            close(er.eps[num].ch)
    close(er.eps[num].ch)
```

Reply:
There is only one call to `copyData` per `num`.

Reply:
The other place has the exact same problem, so it has still the possibility for double close.

Reply:
I'll look again tomorrow with fresh eyes, maybe you are right.

Reply:
But let me try using `CompareAndSwapInt32` instead.